### DISCUSSION TITLE
Visual artifacts with small sprites

### DISCUSSION THREAD

**Discussion by OOPS-Studio** [pixijs/pixijs#8266] [Q&A] [answered]
![image](https://user-images.githubusercontent.com/46879790/161681720-e0d05f19-acf1-407d-abf4-7ea1c3ee80f5.png)

You can see in this image that there are extra partial-pixels being rendered underneath the sprites on the third row.

All of these textures are in the same spritesheet. These are being rendered as `AnimatedSprite`s with the data being passed in directly from the spritesheet loader.

The visual artifacts underneath the sprites are actually showing other parts of the spritesheet. The pixels that show there align perfectly with the other pixels in the spritesheet below that sprite's texture, so it's somehow taking too much data from the spritesheet and rendering a very small sliver of the extra data.

I am putting a container onto the canvas and scaling that container by 2.5x3 to fit the canvas, however even when I don't scale the container at all or use different values for the scale this artifact still shows up unaffected.

Also moving around the sprites does fix the issue, but only when they're in certain positions. About 50/50 chance that any given position will cause the issue vs fix the issue.

It doesn't always happen to the third row of sprites. If I offset all the sprites by one pixel from the top then it happens on the 4th row. 2 pixels from top makes it happen on 5th and 1st. Various different adjustments cause it to happen in various different places, but as far as I can tell it's only on the Y axis.

I can't get the issue to recreate itself on the X axis, and moving them horizontally doesn't seem to have any effect.

I just found out this same bug applies even when I do not enable nearest-neighbor scaling.

This is currently my full source code:
```
import * as PIXI from "pixi.js";

const app = new PIXI.Application({
	view: document.getElementById("pixi-canvas") as HTMLCanvasElement,
	resolution: window.devicePixelRatio || 1,
	autoDensity: true,
	backgroundColor: 0x6495ed,
	width: 640,
	height: 672
});
// 256 x 224

enum ALIEN_TYPES{
	ONE,
	TWO,
	THREE
}

const mainContainer = new PIXI.Container();
mainContainer.scale.set(2.5, 3);
app.stage.addChild(mainContainer);

class Alien extends PIXI.AnimatedSprite{
	type: ALIEN_TYPES;

	constructor(type: ALIEN_TYPES,position: PIXI.Point) {
		const animations = app.loader.resources["spritesheet.json"].spritesheet!.animations;
		let animation: PIXI.Texture[];

		switch (type) {
			case ALIEN_TYPES.ONE:
				animation = animations["alien1"];
				break;
			case ALIEN_TYPES.TWO:
				animation = animations["alien2"];
				break;
			case ALIEN_TYPES.THREE:
				animation = animations["alien3"];
				break;
		}

		super(animation);

		this.position.set(position.x,position.y);

		this.type = type;

		this.init();
	}

	init(): void {
		this.animationSpeed = 0.01;

		this.play();
	}
}

class AlienSwarm extends PIXI.Container{
	constructor() {
		super();

		this.init();
	}

	init(): void {
		for (let i = 0; i < 5; i++){
			for (let j = 0; j < 11; j++){
				const position = new PIXI.Point(j * 15, i * 13);
				switch (i) {
					case 0:
						this.addChild(new Alien(ALIEN_TYPES.THREE,position));
						break;
					case 1:
					case 2:
						this.addChild(new Alien(ALIEN_TYPES.TWO,position));
						break;
					case 3:
					case 4:
						this.addChild(new Alien(ALIEN_TYPES.ONE,position));
						break;
				}
			}
		}
	}
	update(delta: number): void {
		delta// Currently this is just here to prevent TSC from throwing a "delta declared but never read" error
	}
}

PIXI.settings.SCALE_MODE = PIXI.SCALE_MODES.NEAREST;

app.loader.add("spritesheet.json").load(setup);

let alienSwarm: AlienSwarm;

function setup(): void {
        alienSwarm = new AlienSwarm();
	
	mainContainer.addChild(alienSwarm);

	app.ticker.add((delta): void => {
		alienSwarm.update(delta);
	});
}
```

Here is my spritesheet.json
```
{
	"frames": {
		"alien1-1.png": {
			"frame": {"x":0,"y":0,"w":12,"h":8}
		},
		"alien1-2.png": {
			"frame": {"x":36,"y":0,"w":12,"h":8}
		},
		"alien2-1.png": {
			"frame": {"x":12,"y":0,"w":12,"h":8}
		},
		"alien2-2.png": {
			"frame": {"x":48,"y":0,"w":12,"h":8}
		},
		"alien3-1.png": {
			"frame": {"x":24,"y":0,"w":12,"h":8}
		},
		"alien3-2.png": {
			"frame": {"x":60,"y":0,"w":12,"h":8}
		}
	},
	"animations": {
		"alien1": ["alien1-1.png","alien1-2.png"],
		"alien2": ["alien2-1.png","alien2-2.png"],
		"alien3": ["alien3-1.png","alien3-2.png"]
	},
	"meta": {
		"image": "spritesheet.png"
	}
}
```
And here is my spritesheet
![spritesheet](https://user-images.githubusercontent.com/46879790/161819026-b17a89b9-ee17-4f92-a6cf-10205de3c2e4.png)

**Comment by bigtimebuddy** [comment, marked as answer]
To avoid bleeding of sprites, please make sure to add padding to your sprites of at least 1px.
![Screen Shot 2022-04-05 at 2 25 26 PM](https://user-images.githubusercontent.com/864393/161824328-5b3e9940-8f81-4046-9583-a7f538a9945b.png)

**Comment by OOPS-Studio** [comment]
Is that the only solution to this problem? I'm okay doing that, and it's a very simple and easy and fast solution (thanks for suggesting it!), but why is it necessary to do that?

**Reply by bigtimebuddy** [reply]
This is a common issue with atlases and rendering in general. TexturePacker and other atlas-generating tools often have a padding setting just for this purpose. Generating atlases by hand is not as common, in my experience, except for may demos or experiments. It is best practice for production to introduce padding in your sprite-sheets.

In your example, there are a few things that make the lack of padding especially obvious:

1. You are scaling your container the contains your sprites, which introduces more precision issues
2. You are are using nearest for scaleMode, which makes the problem much more obvious

The reason this happens is because ultimately the texture rects stored in your sprite-sheets (pixel coordinates) are converted into UV (a float value from 0 to 1). It is this precision _plus_ interpolation that produces these artifacts.

I know it seems like a bug, but I can assure you that it's not. If you want to completely avoid this issue, you can also use different images for each texture (at the expense of some network latency cost and maybe de-optimizing Pixi's batching)

**Reply by OOPS-Studio** [reply]
But even when I don't scale the container at all and disable nearest-neighbor scaling it still happens and it's still equally as obvious (and looks exactly the same, just obviously smaller since no scaling)

Also, if I'm understanding correctly, you're saying that the coordinates accessed inside a spritesheet are simply values between 0 and 1? So if I had a spritesheet that was 100 pixels wide with 10 sprites in it, each sprite taking up 10 pixels in width... Then the renderer would look for a sprite at X position 20 by simply saying 0.2 instead? And since 0.2 * 100 = 20 it would find the right sprite?
If so, wouldn't that impose a limit on the size of spritesheets? Computers only have so much floating point precision, so a spritesheet with several million pixels would consistently miss the targeted sprites by several pixels?

If I did understand correctly, then may I ask why it's done this way? And doesn't that mean that in some cases a whole pixel of padding isn't enough, and I'd sometimes need 2 or more pixels worth of padding?

I'll assume it's not a bug, and that does make me feel better; glad it's not a hacky workaround and it's actually a limitation - I'll certainly keep this in mind for the future. Thanks for your help and your explanation!

**Comment by OOPS-Studio** [comment]
Also, why does it only happen on the Y axis? Should I also add padding to my sprites on the X axis as well just in case that starts causing issues too?

**Reply by bigtimebuddy** [reply]
This is not an issue with X-axis because you have 1 pixel space in between the character sprites you referencing.
